File: python/data-clumps/good/customer_service.py
class CustomerService:
    def create_customer(self, person, address):
        if not person.is_valid():
            raise Exception('Invalid person data')

        if not address.is_valid():
            raise Exception('Invalid address data')

        # Create customer record
        customer_data = {
            'first_name': person.get_first_name(),
            'last_name': person.get_last_name(),
            'email': person.get_email(),
            'phone': person.get_phone(),
            'date_of_birth': person.get_date_of_birth(),
            'street': address.get_street(),
            'city': address.get_city(),
            'state': address.get_state(),
            'zip_code': address.get_zip_code()
        }

        # Save to database (simulated)
        return customer_data

    def update_customer_address(self, customer_id, address):
        if not address.is_valid():
            raise Exception('Invalid address data')

        # Update address
        address_data = {
            'customer_id': customer_id,
            'street': address.get_street(),
            'city': address.get_city(),
            'state': address.get_state(),
            'zip_code': address.get_zip_code()
        }

        # Save to database (simulated)
        return address_data

    def send_welcome_email(self, person, address):
        message = f"Welcome {person.get_full_name()}!\n\n"
        message += f"Your address: {address.to_string()}\n"

        # Send email (simulated)
        return {'to': person.get_email(), 'message': message}

    def validate_shipping_address(self, address):
        return address.is_valid()

    def format_address_label(self, person, address):
        return f"{person.get_full_name()}\n{address.to_label_format()}"

File: java/data-clumps/good/CustomerService.java
import java.util.*;

public class CustomerService {
    public Map<String, Object> createCustomer(Person person, Address address) throws Exception {
        if (!person.isValid()) {
            throw new Exception("Invalid person data");
        }

        if (!address.isValid()) {
            throw new Exception("Invalid address data");
        }

        // Create customer record
        Map<String, Object> customerData = new HashMap<>();
        customerData.put("first_name", person.getFirstName());
        customerData.put("last_name", person.getLastName());
        customerData.put("email", person.getEmail());
        customerData.put("phone", person.getPhone());
        customerData.put("date_of_birth", person.getDateOfBirth());
        customerData.put("street", address.getStreet());
        customerData.put("city", address.getCity());
        customerData.put("state", address.getState());
        customerData.put("zip_code", address.getZipCode());

        // Save to database (simulated)
        return customerData;
    }

    public Map<String, Object> updateCustomerAddress(int customerId, Address address) throws Exception {
        if (!address.isValid()) {
            throw new Exception("Invalid address data");
        }

        // Update address
        Map<String, Object> addressData = new HashMap<>();
        addressData.put("customer_id", customerId);
        addressData.put("street", address.getStreet());
        addressData.put("city", address.getCity());
        addressData.put("state", address.getState());
        addressData.put("zip_code", address.getZipCode());

        // Save to database (simulated)
        return addressData;
    }

    public Map<String, String> sendWelcomeEmail(Person person, Address address) {
        String message = "Welcome " + person.getFullName() + "!\n\n";
        message += "Your address: " + address.toString() + "\n";

        // Send email (simulated)
        Map<String, String> result = new HashMap<>();
        result.put("to", person.getEmail());
        result.put("message", message);
        return result;
    }

    public boolean validateShippingAddress(Address address) {
        return address.isValid();
    }

    public String formatAddressLabel(Person person, Address address) {
        return person.getFullName() + "\n" + address.toLabelFormat();
    }
}
